hybrid/compartment_examples/inter_comp_call_secure/main.c

```diff
@@ -0,0 +1,194 @@
+#error "This example only works on CHERI hybrid mode"
+#endif
+
+/*******************************************************************************
+ * Assembly Functions & Consts
+ ******************************************************************************/
+
+extern int switch_compartment();
+extern void comp_f_fn();
+extern void comp_g_fn();
+
+/*******************************************************************************
+ * Types & Consts
+ ******************************************************************************/
+
+const size_t comp_stack_size = 1992;
+const size_t total_comp_size = 5000;
+size_t id = 0;
+
+// Duplicate certain capabilities within the heap of each compartments;
+// currently we save DDC and PCC of `switch_compartment`
+const size_t saved_caps_count = 2;
+void *__capability saved_caps[saved_caps_count];
+
+/* Abstract representation of a compartment. Within 80 bytes we represent:
+ * - an id (8B)
+ * - address to the start of the compartment's stack (8B)
+ *   the size of the stack (8B)
+ * - address to the start of the compartment's heap (8B)
+ *   the size of the heap (8B)
+ * - alignment padding (8B)
+ * - the ddc corresponding to the compartment (16B)
+ * - the function within the compartment to be executed upon switching (16B)
+ */
+struct comp
+{
+	size_t id;
+	void *stack_addr;
+	size_t stack_len;
+	void *heap_addr;
+	size_t heap_len;
+	void *__capability ddc;
+	void *__capability comp_fn;
+};
+
+// ASM offsets, included here for validation
+#include "main.h"
+
+static_assert(COMP_SIZE == sizeof(struct comp), "Invalid `COMP_SIZE` provided");
+static_assert(COMP_OFFSET_STK_ADDR == offsetof(struct comp, stack_addr),
+			  "Invalid `COMP_OFFSET_STK_ADDR` provided.");
+static_assert(COMP_OFFSET_STK_LEN == offsetof(struct comp, stack_len),
+			  "Invalid `COMP_OFFSET_STK_LEN` provided.");
+static_assert(COMP_OFFSET_DDC == offsetof(struct comp, ddc), "Invalid `COMP_OFFSET_DDC` provided.");
+static_assert(COMP_OFFSET_PCC == offsetof(struct comp, comp_fn),
+			  "Invalid `COMP_OFFSET_PCC` provided.");
+
+struct comp comps[COMP_COUNT];
+
+/*******************************************************************************
+ * Privileged Functions
+ ******************************************************************************/
+
+void init_comps()
+{
+	void *__capability switch_cap = (void *__capability) switch_compartment;
+	switch_cap = cheri_bounds_set(switch_cap, 80 * 4);
+	asm("seal %w0, %w0, rb" : "+r"(switch_cap) :);
+	saved_caps[0] = switch_cap;
+
+	void *__capability comps_addr = (void *__capability) &comps;
+	comps_addr = cheri_bounds_set(comps_addr, COMP_COUNT * COMP_SIZE);
+	saved_caps[1] = comps_addr;
+}
+
+/* This function stores the compartment data in memory, so that it can be
+ * accessed by the compartments as needed.
+ */
+void executive_switch(struct comp c)
+{
+	void *__capability comps_addr = saved_caps[1];
+
+	asm("mov c19, %w0\n\t"
+		"mov x0, #0\n\t"
```

Comment:
You need to make sure (basically) that the C compiler doesn't mess with those registers between setting them, and reading them in `switch_compartment`. The only safe way to do that is to make sure that there is no C code in between, which mean writing this whole function in assembly.

Reply:
Since the only thing following is the call to `switch_compartment`, would it be safe to simply include that function call as part of the inline assembly, maybe?

Reply:
You can, but you'll have to mark all potentially-clobbered registers as such, and there are a lot of them. (All the caller-saved registers, basically.) It can be done if that's what you need to do, but it's often easier to put the whole lot into a `.s` (or `.S`) file.

Reply:
I see. So the naive solution then is to move `executive_switch` in our `shared.S` file, or is that still insufficient?